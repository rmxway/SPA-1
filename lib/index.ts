export * from './jest-template';
export * from './template';
