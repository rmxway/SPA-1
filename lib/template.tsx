'use client';

import { memo, ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import { Navbar } from '@/components';
import { TopBlock } from '@/components/ui';
import { checkVersion } from '@/services/localStore';
import { store } from '@/store';
import { ReduxProvider } from '@/store/provider';
import { defaultTheme, GlobalStyles } from '@/theme';

export type TemplateProps = {
	children?: ReactNode | undefined;
};

checkVersion(store);

export const Template = memo(({ children }: TemplateProps) => (
	<ReduxProvider>
		<ThemeProvider theme={defaultTheme}>
			<GlobalStyles />
			<Navbar />
			<TopBlock />
			{children}
		</ThemeProvider>
	</ReduxProvider>
));

export default Template;
