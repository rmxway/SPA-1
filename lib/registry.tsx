'use client';

import { useServerInsertedHTML } from 'next/navigation';
import React, { useState } from 'react';
import { ServerStyleSheet, StyleSheetManager } from 'styled-components';

import { JestTemplate, Template } from '@/lib';

interface StyledRegistryProps {
	isJest?: boolean;
	children: React.ReactNode;
}

export function StyledComponentsRegistry({ isJest, children }: StyledRegistryProps) {
	const [sheet] = useState(() => new ServerStyleSheet());

	const Tag = isJest ? JestTemplate : Template;

	try {
		useServerInsertedHTML(() => {
			const styles = sheet.getStyleElement();
			sheet.instance.clearTag();

			// eslint-disable-next-line react/jsx-no-useless-fragment
			return <>{styles}</>;
		});

		if (typeof window !== 'undefined') return <Tag>{children}</Tag>;

		return (
			<StyleSheetManager sheet={sheet.instance}>
				<Tag>{children}</Tag>
			</StyleSheetManager>
		);
	} catch (error) {
		throw new Error((error as Error).message);
	} finally {
		sheet.instance.clearTag();
	}
}

export default StyledComponentsRegistry;
