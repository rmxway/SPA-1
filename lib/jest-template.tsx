import { ThemeProvider } from 'styled-components';

import { ReduxProvider } from '@/store/provider';
import { defaultTheme, GlobalStyles } from '@/theme';

import { TemplateProps } from './template';

export const JestTemplate = ({ children }: TemplateProps) => (
	<ReduxProvider>
		<ThemeProvider theme={defaultTheme}>
			<GlobalStyles />
			{children}
		</ThemeProvider>
	</ReduxProvider>
);

export default JestTemplate;
