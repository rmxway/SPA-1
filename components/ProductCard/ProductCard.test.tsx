import { render } from '@testing-library/react';

import { StyledComponentsRegistry } from '@/lib/registry';

import { ProductCard } from '.';

const mockProduct = {
	id: 6,
	title: 'Calvin Klein CK One',
	description:
		"CK One by Calvin Klein is a classic unisex fragrance, known for its fresh and clean scent. It's a versatile fragrance suitable for everyday wear.",
	category: 'fragrances',
	price: 49.99,
	discountPercentage: 0.32,
	rating: 4.85,
	tags: ['fragrances', 'perfumes'],
	brand: 'Calvin Klein',
	images: [
		'https://cdn.dummyjson.com/products/images/fragrances/Calvin%20Klein%20CK%20One/1.png',
		'https://cdn.dummyjson.com/products/images/fragrances/Calvin%20Klein%20CK%20One/2.png',
		'https://cdn.dummyjson.com/products/images/fragrances/Calvin%20Klein%20CK%20One/3.png',
	],
	thumbnail: 'https://cdn.dummyjson.com/products/images/fragrances/Calvin%20Klein%20CK%20One/thumbnail.png',
};

describe('ProductCard:', () => {
	it('Render card', () => {
		const { container, getByText } = render(
			<StyledComponentsRegistry isJest>
				<ProductCard product={mockProduct} />
			</StyledComponentsRegistry>,
		);

		const text = getByText(/Calvin Klein CK One/i);
		expect(text).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
});
