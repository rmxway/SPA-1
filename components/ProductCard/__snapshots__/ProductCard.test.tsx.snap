// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard: Render card 1`] = `
<div>
  <div
    class="sc-aa1a80d5-2 bWWIfb"
  >
    <div
      class="sc-7d6be7a8-0 kzlrqJ"
    >
      -
      0
      %
    </div>
    <div
      class="sc-d8b47b8e-0 gDyAOM"
    >
      <i
        class="icofont icofont-favorite-fill "
      />
    </div>
    <a
      href="/product/6"
    >
      <div
        class="sc-aa1a80d5-1 fVfJIC"
        id="wrapper-6"
      >
        <div
          class="sc-91434ab0-0 NzyZw"
        >
          <div
            class="sc-91434ab0-1 kzBAtY"
          />
        </div>
      </div>
    </a>
    <a
      class="sc-aa1a80d5-3 eofkaM"
      href="/product/6"
    >
      <div>
        Calvin Klein CK One
      </div>
      <span>
        Calvin Klein
      </span>
    </a>
    <span
      class="sc-a1f6370b-0 goVLkk"
    />
    <div
      class="sc-aa1a80d5-5 gwhwNr"
    >
      <div
        class="sc-be016b9b-0 hGlQTV"
      >
        <div
          class="sc-be016b9b-1 fHjwaS"
        >
          <div
            class="sc-be016b9b-2 fMxAWE"
          >
            <i
              class="icofont icofont-star sc-be016b9b-3 fICwyA"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fICwyA"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fICwyA"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fICwyA"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fICwyA"
            />
          </div>
          <div
            class="sc-be016b9b-2 etYehl"
          >
            <i
              class="icofont icofont-star sc-be016b9b-3 fnkWPB"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fnkWPB"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fnkWPB"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fnkWPB"
            />
            <i
              class="icofont icofont-star sc-be016b9b-3 fnkWPB"
            />
          </div>
        </div>
        <span>
          4.85
        </span>
      </div>
      <div
        class="sc-aa1a80d5-4 jSVhXR"
      >
        49.99
         
        $
      </div>
    </div>
    <button
      class="sc-6ffd862-0 bkeQbQ"
    >
      Add to cart
      <i
        class="icofont icofont-cart "
      />
    </button>
  </div>
</div>
`;
